eng/native/detect-compiler.sh

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env bash
```

Comment:
nit: we can call it `init-compiler.sh`, to rhyme with the existing `init-distro-rid.sh` and upcoming `init-os-and-arch.sh` (from https://github.com/dotnet/runtime/pull/31701)?

Reply:
yup sounds good